cpp/src/arrow/adapters/orc/adapter_util.cc

```diff
@@ -315,13 +344,662 @@ Status AppendBatch(const liborc::Type* type, liborc::ColumnVectorBatch* batch,
+          array, orc_offset, column_vector_batch);
+    case arrow::Type::type::FIXED_SIZE_BINARY:
+      return WriteFixedSizeBinaryBatch(array, orc_offset, column_vector_batch);
+    case arrow::Type::type::DATE32:
+      return WriteGenericBatch<arrow::Date32Type, liborc::LongVectorBatch>(
+          array, orc_offset, column_vector_batch);
+    case arrow::Type::type::DATE64:
+      return WriteTimestampBatch<arrow::Date64Type>(
+          array, orc_offset, column_vector_batch, kOneSecondMillis, kOneMilliNanos);
+    case arrow::Type::type::TIMESTAMP: {
+      switch (arrow::internal::checked_pointer_cast<arrow::TimestampType>(array.type())
+                  ->unit()) {
+        case arrow::TimeUnit::type::SECOND:
+          return WriteTimestampBatch<arrow::TimestampType>(
+              array, orc_offset, column_vector_batch, 1, kOneSecondNanos);
+        case arrow::TimeUnit::type::MILLI:
+          return WriteTimestampBatch<arrow::TimestampType>(
+              array, orc_offset, column_vector_batch, kOneSecondMillis, kOneMilliNanos);
+        case arrow::TimeUnit::type::MICRO:
+          return WriteTimestampBatch<arrow::TimestampType>(
+              array, orc_offset, column_vector_batch, kOneSecondMicros, kOneMicroNanos);
+        case arrow::TimeUnit::type::NANO:
+          return WriteTimestampBatch<arrow::TimestampType>(
+              array, orc_offset, column_vector_batch, kOneSecondNanos, 1);
+        default:
+          return arrow::Status::Invalid("Unknown or unsupported Arrow type: ",
+                                        array.type()->ToString());
+      }
+    }
+    case arrow::Type::type::DECIMAL128: {
+      int32_t precision =
+          std::static_pointer_cast<arrow::Decimal128Type>(array.type())->precision();
+      if (precision > 18) {
+        return WriteGenericBatch<arrow::Decimal128Type, liborc::Decimal128VectorBatch>(
+            array, orc_offset, column_vector_batch);
+      } else {
+        return WriteGenericBatch<arrow::Decimal128Type, liborc::Decimal64VectorBatch>(
+            array, orc_offset, column_vector_batch);
+      }
+    }
+    case arrow::Type::type::STRUCT:
+      return WriteStructBatch(array, orc_offset, column_vector_batch, normalized);
+    case arrow::Type::type::LIST:
+      return WriteListBatch<arrow::ListArray>(array, orc_offset, column_vector_batch);
+    case arrow::Type::type::LARGE_LIST:
+      return WriteListBatch<arrow::LargeListArray>(array, orc_offset,
+                                                   column_vector_batch);
+    case arrow::Type::type::FIXED_SIZE_LIST:
+      return WriteListBatch<arrow::FixedSizeListArray>(array, orc_offset,
+                                                       column_vector_batch);
+    case arrow::Type::type::MAP:
+      return WriteMapBatch(array, orc_offset, column_vector_batch);
+    default: {
+      return arrow::Status::Invalid("Unknown or unsupported Arrow type: ",
```

Comment:
Please raise `Status::TypeError` or `Status::NotImplemented` here.

Reply:
Thanks! Switched to `Status::NotImplemented` since we don't have dictionary or union support in this PR but they will be supported.